### DISCUSSION TITLE
Plain load balancing strategy

### DISCUSSION THREAD

**Discussion by Devocub** [DNSCrypt/dnscrypt-proxy#2219] [Ideas]
Each DNS request: use first server from server_names, if not available (no response for _configurable_ seconds), use second one, ..., third, ...
Two variants: not sorted and sorted.
As optimization unavailable server can be temporarily (_configurable_ seconds) removed from list.

**Comment by t0ny54** [comment]
i like this:

prefered_server_name = ['server1', 'server2', 'server3', 'server4']   # your favorite dns server list
alternate_server_name = ['server5', 'server6', 'server7', 'server8']  # backup dns server if prefered server name don't have server live

Each prefered or alternate server name, combine with lb_strategy and Sorted latencies.
and only use prefered server name if prefered_server_name have server live... 😀

**Comment by awebeer256** [comment]
Here's my preferred implementation.
Config:
`server_names = <servconf>`
where:
```
<servconf> ::= <servlist> | <confobj>
<servlist> ::= array of strings
<confobj> ::= { main=<servlist>, backup=<servconf> }
```

So we could still write `server_names = ['server1', 'server2']`, but we could also write:
```
server_names = {
  main=['server1', 'server2'],
  backup={
    main=['server3', 'server4'],
    backup={
      main=['server5', 'server6'],
      backup=['server7', 'server8']
    }
  }
}
```
These two config entries would behave identically, except when `server1` and `server2` are both down, in which case the latter would behave identically to `server_names = ['server3', 'server4']`, unless those were _also_ both down, in which case it would behave identically to `server_names = ['server5', 'server6']`, etc.

My particular use case is as follows:

- I want to use DNS filtering
- I want to outsource my choice of filters to the people running the upstream resolvers
- I want to configure dnscrypt-proxy with several upstream resolvers in case some go down
- Each organization that operates a resolver with filtering will use a different selection of filters
- I like some selections more than others
- Each organization only operates a small number of resolvers

So you can see why I might want to tell dnscrypt-proxy, "Always use resolver X if it's up, otherwise do your regular load-balancing among Y and Z, and if all of those are down use W instead." Right now, I have to use `lb_strategy = 'first'`, and choose X, Y, Z, and W and their relays such that X will (almost) always be faster than Y or Z, both of which will always be faster than Z.
